### DISCUSSION TITLE
Bruggeman EMA returns unexpected results

### DISCUSSION THREAD

**Discussion by mrtnrb** [PyEllips/pyElli#171] [Q&A] [answered]
Hello,

I receive unexpected results in specific cases with the Bruggemann EMA. I want to model some kind of roughness, means I use air as the host material and a metal as the guest. The returned material however has negative extinction coefficient. It seems to me that in these cases the unphysical root of the BEMA equation is chosen. The code illustrating this behavior is here:

```
import elli
from elli.materials import BruggemanEMA as BEMA
import matplotlib.pyplot as plt
from numpy import sqrt
import numpy as np
rii_db = elli.db.RII()
#%%

def BEMAroots(host_material, guest_material, lbda, fraction):
    """Gets the permittivity tensor of the BEMA material for wavelength 'lbda',
    """
    e_h = host_material.get_tensor(lbda)
    e_g = guest_material.get_tensor(lbda)
    f = fraction
    root1 = 3*e_g*f/4 - e_g/4 - 3*e_h*f/4 + e_h/2 - sqrt(
        9*e_g**2*f**2 - 6*e_g**2*f + e_g**2 - 18*e_g*e_h*f**2 +
        18*e_g*e_h*f + 4*e_g*e_h + 9*e_h**2*f**2 - 12*e_h**2*f + 4*e_h**2)/4
    root2 = 3*e_g*f/4 - e_g/4 - 3*e_h*f/4 + e_h/2 + sqrt(
        9*e_g**2*f**2 - 6*e_g**2*f + e_g**2 - 18*e_g*e_h*f**2 +
        18*e_g*e_h*f + 4*e_g*e_h + 9*e_h**2*f**2 - 12*e_h**2*f + 4*e_h**2)/4

    return root1, root2

#%% 
def plotBEMA(book, page, wvl, fc):
    """plots n,k resulting from Bruggemann EMA model for material retrieved 
    from refractiveindex.info: pyelli result, root1, root2, and original 
    material
    """
    mat = rii_db.get_mat(book, page)
    
    B1 = BEMA(elli.AIR, mat, fc)
    N = np.array([np.unique(Nar[Nar.nonzero()]) 
                  for Nar in B1.get_refractive_index(wvl)])
    r1, r2 = BEMAroots(elli.AIR, mat, wvl, fc)
    Nr1 = sqrt(np.array([np.unique(r[r.nonzero()]) for r in r1]))
    Nr2 = sqrt(np.array([np.unique(r[r.nonzero()]) for r in r2]))
    
    Nme = np.array([np.unique(Nar[Nar.nonzero()]) 
                              for Nar in mat.get_refractive_index(wvl)])
    plt.figure()
    ax1=plt.axes()
    
    ax1.plot(wvl, np.real(N), color = 'Tab:blue',)
    ax1.plot(wvl, np.imag(N),'--', color = 'Tab:blue',)
    
    ax1.plot(wvl, np.real(Nr1),  'x', color = 'black')
    ax1.plot(wvl, np.imag(Nr1), 'o', color = 'black')
    
    ax1.plot(wvl, np.real(Nr2),  'x', color = 'Tab:red')
    ax1.plot(wvl, np.imag(Nr2), 'o', color = 'Tab:red')
    
    ax1.plot(wvl, np.real(Nme),  color = 'lightgrey')
    ax1.plot(wvl, np.imag(Nme), '--', color = 'lightgrey')
    
    ax1.set_title('{} {} with f = {}'.format(book, page, fc))
    
    ax1.set_xlabel(r'$\lambda$ / nm')
    ax1.legend(['n (Pyelli)','k(Pyelli)', 
                'n(root1)', 'k(root1)', 
                'n(root2)', 'k(root2)',
                'n(Mat)', 'k(Mat)'])

#%% Plot a few examples:
wvl = np.arange(400,725,25)

plotBEMA("Au", "Johnson", wvl, 0.1)
plotBEMA("Au", "Johnson", wvl, 0.5)
plotBEMA("Rh", "Arndt", wvl, 0.1)
plotBEMA("Rh", "Arndt", wvl, 0.5)
plotBEMA("Si", "Aspnes", wvl, 0.1)
plotBEMA("Si", "Aspnes", wvl, 0.5)
```

As examples I have chosen two metals: Au as a standard and Rh as this comes close to my use case. For both, negative k is returned by pyelli, although I would choose the other root as the "physical" solution. Also for Au f=0.1 one can see that sometimes the returned solution jumps between root1 and root2, which seems reasonable due to  consistency but still seems to be the "unphysical" result. Interestingly, for Si the behaviour is as I would expect it. Is this a problem in the algorithm or is there a generel problem in using the BEMA with metals?

Best,
Martin

**Comment by MarJMue** [comment]
Hey Martin,

thank you for the detailed report. I will have a look and try to find out what is going wrong. It may not be the most stable implementation to find the correct root. 

It could be helpful if you can tell me which version of Numpy you are using.

I will come back to you when i know more.

Best
Marius

**Reply by mrtnrb** [reply]
It is numpy version 1.26.0
M

**Reply by MarJMue** [reply]
I have found one error in our code, which fixed the Rhodium and Silicon examples. The Gold still does not work.

While reading up on the matter, I have found an alternative solution, which handles all your examples correctly. 
I have implemented it in PR #173, if you want to, you can test the code from there, as I want to wait with releasing the fix until I have run a few more tests.

**Reply by mrtnrb** [reply]
I tried it for my my use case which is fitting a thin layer of metal. In several variations the new algorithm always used the correct root. Thanks a lot.

**Reply by MarJMue** [reply, marked as answer]
To ping you as well. We have released a new version of pyelli, which includes the fixed Berreman EMA.
Thank you very much for pointing out the issue.
